src/Npgsql/NpgsqlConnector.cs

```diff
@@ -1013,6 +1013,10 @@ internal ValueTask<IBackendMessage> ReadMessage(bool async, DataRowLoadingMode d
                 return ReadMessageLong(dataRowLoadingMode, readingNotifications2: false);
             }
 
+            // Just in case if we were successful in sending a cancellation, but the query is already completed (and the response is already in the buffer)
```

Comment:
This is a good catch! However... This method is an (extreme) optimization for the case where all the data for the next backend message is already buffered in our internal buffer; the idea is that we don't have to do any I/O at all (everything's already in memory), so we don't need to pay the overhead of an async state machine, etc. That's why it calls into ReadMessageLong the moment it detects missing data, or weird cases which are a bit more complicated.

Long story short, I'd move the resetting of originalTimeoutException into ParseServerMessage for ReadyForQuery, that ensures it will always get called.

Reply:
Okay, this doesn't work. The problem is, before the `RFQ` we get an error from PG (about cancelling the request), and set this error into a variable. Then, we read the `RFQ` and if there was the error, we throw it (without resetting the original timeout exception). 

So, for this to work inside of the ParseServerMessage, it should be aware of any error from a previous cycle. For now, I revert this change.

Reply:
Makes sense. The code as it is now looks good to me.